provider/azure/internal/azureauth/interactive.go

```diff
@@ -206,17 +209,44 @@ func createOrUpdateServicePrincipal(
 		return "", "", errors.Annotate(err, "preparing password credential")
 	}
 
-	servicePrincipal, err := client.Create(
-		ad.ServicePrincipalCreateParameters{
-			ApplicationID:       jujuApplicationId,
-			AccountEnabled:      true,
-			PasswordCredentials: []ad.PasswordCredential{passwordCredential},
+	// Attempt to create the service principal. When the user
+	// authenticates, Azure will replicate the application
+	// into the user's AAD. This happens asynchronously, so
+	// it may not exist by the time we try to create the
+	// service principal; thus, we retry until it exists. The
+	// error checking is based on the logic in azure-cli's
+	// create_service_principal_for_rbac.
+	var servicePrincipal ad.ServicePrincipal
+	createServicePrincipal := func() error {
+		var err error
+		servicePrincipal, err = client.Create(
+			ad.ServicePrincipalCreateParameters{
+				ApplicationID:       jujuApplicationId,
+				AccountEnabled:      true,
+				PasswordCredentials: []ad.PasswordCredential{passwordCredential},
+			},
+			nil, // abort
+		)
+		return err
+	}
+	retryArgs := retry.CallArgs{
+		Func: createServicePrincipal,
+		IsFatalError: func(err error) bool {
+			serviceErr, ok := errorutils.ServiceError(err)
+			if ok && (strings.Contains(serviceErr.Message, " does not reference ") ||
+				strings.Contains(serviceErr.Message, " does not exist ")) {
+				// The application doesn't exist yet, retry later.
+				return false
+			}
+			return true
 		},
-		nil, // abort
-	)
-	if err != nil {
+		Clock:       clock,
+		Delay:       time.Second,
```

Comment:
Actually I'll just make it 5 seconds, which is what azure-cli does. Should be fine.